File: src/org/antlr/runtime/tree/BaseTreeAdaptor.java
package org.antlr.runtime.tree;

import org.antlr.runtime.Token;

public abstract class BaseTreeAdaptor implements TreeAdaptor {
	public Object nil() {
		return create(null);
	}

	public Object dupTree(Object tree) {
		return ((Tree)tree).dupTree();
	}

	/** Add a child to the tree t.  If child is a flat tree (a list), make all
	 *  in list children of t.
	 */
	public void addChild(Object t, Object child) {
		((Tree)t).addChild((Tree)child);
	}

	/** If oldRoot is a nil root, just copy or move the children to newRoot.
	 *  If not a nil root, make oldRoot a child of newRoot.
	 *
	 *    old=^(nil a b c), new=r yields ^(r a b c)
	 *    old=^(a b c), new=r yields ^(r ^(a b c))
	 *
	 *  If newRoot is a nil-rooted single child tree, use the single
	 *  child as the new root node.
	 *
	 *    old=^(nil a b c), new=^(nil r) yields ^(r a b c)
	 *    old=^(a b c), new=^(nil r) yields ^(r ^(a b c))
	 *
	 *  If oldRoot was null, it's ok, just return newRoot (even if isNil).
	 *
	 *    old=null, new=r yields r
	 *    old=null, new=^(nil r) yields ^(nil r)
	 *
	 *  Return newRoot.  Throw an exception if newRoot is not a
	 *  simple node or nil root with a single child node--it must be a root
	 *  node.
	 *
	 *  Be advised that it's ok for newRoot to point at oldRoot's
	 *  children; i.e., you don't have to copy the list.  We are
	 *  constructing these nodes so we should have this control for
	 *  efficiency.
	 */
	public Object becomeRoot(Object newRoot, Object oldRoot) {
		Tree newRootTree = (Tree)newRoot;
		Tree oldRootTree = (Tree)oldRoot;
		if ( oldRoot==null ) {
			return newRoot;
		}
		// handle ^(nil real-node)
		if ( newRootTree.isNil() ) {
			if ( newRootTree.getChildCount()>1 ) {
				// TODO: make tree run time exceptions hierarchy
				throw new RuntimeException("more than one node as root (TODO: make exception hierarchy)");
			}
			newRootTree = (Tree)newRootTree.getChild(0);
		}
		// add oldRoot to newRoot; addChild takes care of case where oldRoot
		// is a flat list (i.e., nil-rooted tree).  All children of oldRoot
		// are added to newRoot.
		newRootTree.addChild(oldRootTree);
		return newRootTree;
	}

	/** Transform ^(nil x) to x */
	public Object rulePostProcessing(Object root) {
		Tree r = (Tree)root;
		if ( r!=null && r.isNil() && r.getChildCount()==1 ) {
			r = (Tree)r.getChild(0);
		}
		return r;
	}

	public void addChild(Object t, Token child) {
		addChild(t, create(child));
	}

	public Object becomeRoot(Token newRoot, Object oldRoot) {
		return becomeRoot(create(newRoot), oldRoot);
	}

	public Object create(int tokenType, Token fromToken) {
		fromToken = createToken(fromToken);
		//((ClassicToken)fromToken).setType(tokenType);
		fromToken.setType(tokenType);
		Tree t = (Tree)create(fromToken);
		return t;
	}

	public Object create(int tokenType, Token fromToken, String text) {
		fromToken = createToken(fromToken);
		fromToken.setType(tokenType);
		fromToken.setText(text);
		Tree t = (Tree)create(fromToken);
		return t;
	}

	public Object create(int tokenType, String text) {
		Token fromToken = createToken(tokenType, text);
		Tree t = (Tree)create(fromToken);
		return t;
	}

	public int getType(Object t) {
		((Tree)t).getType();
		return 0;
	}

	public void setType(Object t, int type) {
	}

	public void setText(Object t, String text) {
	}

	public Object getChild(int i) {
		return null;
	}

	public int getChildCount() {
		return 0;
	}
}

File: runtime/CSharp/Antlr3.Runtime/Antlr.Runtime.Tree/BaseTreeAdaptor.cs
namespace Antlr.Runtime.Tree
{
    using Antlr.Runtime;
    using System;
    using System.Runtime.CompilerServices;

    public abstract class BaseTreeAdaptor : ITreeAdaptor
    {
        protected BaseTreeAdaptor()
        {
        }

        public virtual void AddChild(object t, object child)
        {
            if (t != null)
            {
                ((ITree) t).AddChild((ITree) child);
            }
        }

        public virtual object BecomeRoot(IToken newRoot, object oldRoot)
        {
            return this.BecomeRoot(this.Create(newRoot), oldRoot);
        }

        public virtual object BecomeRoot(object newRoot, object oldRoot)
        {
            ITree child = (ITree) newRoot;
            ITree t = (ITree) oldRoot;
            if (oldRoot == null)
            {
                return newRoot;
            }
            if (child.IsNil)
            {
                if (child.ChildCount > 1)
                {
                    throw new SystemException("more than one node as root (TODO: make exception hierarchy)");
                }
                child = child.GetChild(0);
            }
            child.AddChild(t);
            return child;
        }

        public abstract object Create(IToken param1);
        public virtual object Create(int tokenType, IToken fromToken)
        {
            fromToken = this.CreateToken(fromToken);
            fromToken.Type = tokenType;
            return (ITree) this.Create(fromToken);
        }

        public virtual object Create(int tokenType, string text)
        {
            IToken token = this.CreateToken(tokenType, text);
            return (ITree) this.Create(token);
        }

        public virtual object Create(int tokenType, IToken fromToken, string text)
        {
            fromToken = this.CreateToken(fromToken);
            fromToken.Type = tokenType;
            fromToken.Text = text;
            return (ITree) this.Create(fromToken);
        }

        public abstract IToken CreateToken(IToken fromToken);
        public abstract IToken CreateToken(int tokenType, string text);
        public abstract object DupNode(object param1);
        public virtual object DupTree(object tree)
        {
            return ((ITree) tree).DupTree();
        }

        public virtual object GetChild(object t, int i)
        {
            return ((ITree) t).GetChild(i);
        }

        public virtual int GetChildCount(object t)
        {
            return ((ITree) t).ChildCount;
        }

        public virtual object GetNilNode()
        {
            return this.Create(null);
        }

        public virtual string GetNodeText(object t)
        {
            return ((ITree) t).Text;
        }

        public virtual int GetNodeType(object t)
        {
            int type = ((ITree) t).Type;
            return 0;
        }

        public abstract IToken GetToken(object treeNode);
        public abstract int GetTokenStartIndex(object t);
        public abstract int GetTokenStopIndex(object t);
        public int GetUniqueID(object node)
        {
            return RuntimeHelpers.GetHashCode(node);
        }

        public virtual bool IsNil(object tree)
        {
            return ((ITree) tree).IsNil;
        }

        public virtual object RulePostProcessing(object root)
        {
            ITree child = (ITree) root;
            if (((child != null) && child.IsNil) && (child.ChildCount == 1))
            {
                child = child.GetChild(0);
            }
            return child;
        }

        public virtual void SetNodeText(object t, string text)
        {
            throw new NotImplementedException("don't know enough about Tree node");
        }

        public virtual void SetNodeType(object t, int type)
        {
            throw new NotImplementedException("don't know enough about Tree node");
        }

        public abstract void SetTokenBoundaries(object param1, IToken param2, IToken param3);
    }
}

